### ISSUE TITLE
Some buckets are missing

### ISSUE THREAD

**Issue by khelle** [stellar/stellar-core#1518]
I have another problem with Stellar Core. After resolving the problems with joining the public network, I was running the node for few days without problems. The start/restart worked flawlessly, the node connected each time to network, but now I encountered another problem. Despite node claiming to be fully synced, it does not see any transaction made. It did for some time as I tested it via Horizon API connected to the node, but now it does not see anything. When I look at the logs I can see:

```
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.334 GBA2J [default INFO] Starting stellar-core v9.0.1-6-g226ba55
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.334 GBA2J [default INFO] Config from /home/xlm/.stellar-core/stellar-core.cfg
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.335 GBA2J [Database INFO] Connecting to: postgresql://dbname=dbname user=user password=******** host=127.0.0.1
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.350 GBA2J [SCP INFO] LocalNode::LocalNode@GBA2J qSet: b3a703
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.350 GBA2J [default INFO] Listening on 127.0.0.1:12059 for HTTP requests
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.352 GBA2J [History INFO] Archive 'local' has 'put' and 'get' commands, will be read and written
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.352 GBA2J [History INFO] Archive 'sdf1' has 'get' command only, will not be written
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.352 GBA2J [History INFO] Archive 'sdf2' has 'get' command only, will not be written
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.352 GBA2J [History INFO] Archive 'sdf3' has 'get' command only, will not be written
Jan 30 12:52:06 localhost stellar-core[11412]: 2018-01-30T12:52:06.353 GBA2J [default INFO] Loading last known ledger
Jan 30 12:52:07 localhost stellar-core[11412]: 2018-01-30T12:52:07.013 GBA2J [Ledger WARNING] Some buckets are missing in '/home/xlm/buckets'.
Jan 30 12:52:07 localhost stellar-core[11412]: 2018-01-30T12:52:07.014 GBA2J [Ledger WARNING] Attempting to recover from the history store.
Jan 30 12:52:07 localhost stellar-core[11412]: 2018-01-30T12:52:07.019 GBA2J [History INFO] Starting RepairMissingBucketsWork
```

After that at some point:

```
stellar-core: bucket/FutureBucket.cpp:267: void stellar::FutureBucket::startMerge(stellar::Application &): Assertion `snap' failed.
Aborted
```

The node then crashes. What's wrong this time? Could the buckets be damaged by running Horizon?

My system specs:

```
Ubuntu 16.04 x64
Stellar-core v9.0.1-6
SSD Hard Drive
Network Connection 100 Mbps
```

Config and SCP problem with resultion: https://github.com/stellar/stellar-core/issues/1487

**Comment by MonsieurNicolas** [comment]
Thanks for reporting this one! Nobody so far reported the "short version" of this.
We've only got reports such as #1296 without clear repro steps.

The fact that you got this to happen by just doing a stop/start cycle seems to indicate that the two systems (SQL and bucket list) got out of sync.

A bit more detail for @vogel .
After looking a bit at the code in [LedgerManagerImpl::closeLedger](https://github.com/stellar/stellar-core/blob/v9.1.0/src/ledger/LedgerManagerImpl.cpp#L684) here is a theory for what is going on (I am using the latest release 9.1.0 but this has been present for ever from what I can tell):
1. a sql transaction is [opened](https://github.com/stellar/stellar-core/blob/v9.1.0/src/ledger/LedgerManagerImpl.cpp#L716)
2. the data is passed down to the bucket subsystem with [ledgerClosed](https://github.com/stellar/stellar-core/blob/v9.1.0/src/ledger/LedgerManagerImpl.cpp#L760)
3. (at least) level 0 is updated by calling `commit`. which in turn calls [setCurr](https://github.com/stellar/stellar-core/blob/v9.1.0/src/bucket/BucketList.cpp#L101) and finally replaces `mCurr` with the new value but while doing so it [marks the old bucket for deletion](https://github.com/stellar/stellar-core/blob/v9.1.0/src/bucket/BucketList.cpp#L73) 
4. the sql transaction is rolled back ; it would have been committed by the [commit](https://github.com/stellar/stellar-core/blob/v9.1.0/src/ledger/LedgerManagerImpl.cpp#L785)

At that point:
* the database points to the old BucketList (as the SQL tx got rolledback)
* the old buckets are (going to be) deleted as the destructor or `Bucket` deletes buckets that don't have the retain flag

Now, how to fix this?

The fix must do a few things:
* add a try/catch block in `LedgerManagerImpl::closeLedger` to call `assumeState` with the value of whatever is in the database after the SQL tx gets rolled back (this avoids stellar-core continuing to run with a bad state, which could cause all sorts of undefined behavior)
* ditch the notion of "retaining" buckets (this was already taken care of by #1459). This ensures that the only place where we delete buckets is in the GC code (executed after the SQL commit); I think this would cause the buckets to be retained even in the case of a crash (let's say it crashes right after updating the bucket list but before the sql tx commit). We need to verify that this is a correct assumption.

**Comment by khelle** [comment]
@MonsieurNicolas I believe the problem happened because of running Horizon. As I said I was using the node without the problem after I resolved problem #1487 , forced it to start / stop between random periodcs of time (always more than 1 min) and the sync contiued without the problem. I succeeded sending and receivng the transaction. Then I decided its time to integrate the XLM into production, started Horizon in ingest mode, and then problems started. Horizon reported me that he cannot find cursor in its database and its probably damaged. Since then core started to report this bucket problems. It can be only a coincidence this happened at same time, but I think its highly possible the two of them cooperated wrongly and one broke something for the other. This is especially viable from my perspective as they seem to oparate and modifiy the same database.

Is there anything I can do, or my only way to solve this now is to remove bucket directory and re-initialize both databases? The core and horizon ones? Can I be given exact steps that would help?

**Comment by MonsieurNicolas** [comment]
I think you were just unlucky on this one (go buy a lottery ticket :) ): Horizon's cursor is independent from the bucket list. Buckets are only used to compute the secure hash of the ledger and the "cursors" are a secondary table independent from everything else really.
I found another report (in a different form) that is alluding to the same problem ( #1395 ).

As to recover from this: sadly the only way right now to do so is to destroy core's state with `newdb` as the buckets got deleted from the local filesystem but are not snapshots (and therefore not available for download from the archive).

**Comment by vogel** [comment]
@MonsieurNicolas I do not think rollback is possible in your scenerio.

Rollback would require an exception between ledgerClosed and txscope.commit(); - that exception would not be caught anywhere and would cause stellar-core to exit.

It would also mean that unreferenced buckets were not removed, as it is done in 'step 4' later.

**Comment by MonsieurNicolas** [comment]
@vogel : re-read the steps I outlined. Yes the exception will pop up all the way up.
In 9.1, the buckets are deleted in the destructor, so an exception should trigger that.

Now, in the new implementation you are correct that it may not be possible anymore as we delete only during the GC step.

**Comment by MonsieurNicolas** [comment]
I am re-opening this issue.

After upgrading a node to 10.0.0, it came back up with:
```
2018-09-12T13:44:33.155 GC5SX [default INFO] Loading last known ledger
2018-09-12T13:44:33.161 GC5SX [Ledger WARNING] Some buckets are missing in '/data/buckets'.
```

Interesting entry in the logs before getting into this state (node crashed because of this which is fine, the problem is the failure mode):
```
2018-09-12T11:10:31.885 GC5SX [Ledger INFO] Closed ledger: [seq=19948007, hash=155198]
terminate called after throwing an instance of 'soci::postgresql_soci_error'
what(): Cannot begin transaction. server closed the connection unexpectedly
This probably means the server terminated abnormally
before or while processing the request.
```

Looks like some changes are done (like deleting buckets?) even when there is a problem with SQL

**Comment by MikeC-BC** [comment]
what's the fix for this @MonsieurNicolas ?
